draft-ietf-ppm-dap.md

```diff
@@ -1115,35 +1292,35 @@ attacks.]
 ~~~
 struct {
   TaskID task_id;
-  Interval batch_interval;
+  Query query;
```

Comment:
> In the fixed chunk model, however, we grant the leader total power to assign reports to chunks (modulo the max and min sizes the task defines), so we should carefully analyze what powers that gives the leader.

Nice catch. There should already be an OPEN ISSUE somewhere in this PR about this.

> For instance, suppose the leader can see the uploading client's IP and do geolocation. Maybe it constructs one chunk that only contains reports from clients in country A and another chunk that only contains reports from country B in an effort to skew either chunk's aggregation somehow?

Agreed, this is a risk worth considering. I imagine there are other things the Leader can do like this already, but you're right that the intention of including the timestamp in the report was to allow the Client to control (at least in part), which batch its report goes in.

It's worth pointing out that replay protection requirements haven't changed: Each Aggregator still needs to ensure that a single report doesn't get replayed in multiple chunks.

> My intuition on how this feature is used is that you want to enable the leader to create a chunk every `min_chunk_size` reports. So maybe we want to enable the client to verify that the chunks defined by the leader are made up of non-overlapping intervals? That would require putting the `Time` component back in a report `Nonce`, but this would use code the helper needs to have anyway to enforce that batch interval queries don't overlap.

Ohhhh that's an interesting idea! Your intuition is right I think. It's a little stricter than what this PR does now. (And thus harder to implement?). The Leader can still mount a similar "attack" by only processing reports from the Client population it's interested in. However there is a cost: It'll end up losing the data for all other Clients.

Feedback on feasibility of the implementation would be useful.

Reply:
I think this concern is valid, but making sure each chunk has non-overlapping intervals can cause us to lose some late-arriving reports. For example, assuming max_batch_lifetime is 1 for simplicity, and I have a task that can quickly fulfill min chunk size in a short period of time (e.g. 1 minute). If leader has constructed a chunk that meets minimum chunk size, the min and max time of all the reports in this chunk defines the interval for this chunk. Now for the next chunk, leader is only allowed to collect reports that have timestamps later than the max timestamp of the previous chunk. If there are late arriving reports, leader is forced to drop all of them even though they are valid, so the availability of the leader service may take a hit in this case.

Reply:
re: Brandon's points:

(1): This was already a case we had to handle in the batch interval query, and I think we would handle it the same way here: if a report arrives after the collect job for its chunk has been serviced, it gets ignored.

(2): Ooh, this is interesting. I think this can be handled, though. Let's say a collector comes along and asks for the result of aggregating chunk ID 4, and then immediately asks for chunk ID 5. The leader will generate a couple of collect job URIs and send those back to the collector -- at this point there's no aggregator commitment to which reports go in which chunk. The leader might then identify a set of candidate reports for chunk 4 and another set for chunk 5. It then carves each set of reports into some number of aggregation jobs and executes them with the helper.

Now, suppose enough reports are rejected from the candidate set assigned to chunk 4 that `min_batch_size` isn't satisfied. The leader will want to then steal some aggregation jobs from chunk 5 to get chunk 4 over the minimum.

OK, but what if the aggregation jobs for chunk 5 finished first and the aggregate result was delivered to the collector? In that case, yes, we are stuck and can never satisfy the collect request for chunk 4. But we could assert that the leader can't deliver the results for the collect job for chunk 5 until it has delivered the results for chunk 4. Note that the constraint here is not on when or in what order the aggregators can do the work of preparation. The constraint is on when the results can be delivered, because it's only at that point that the aggregators commit to assigning a set of reports to a chunk.

So I think this means that aggregators could still be running all the aggregate jobs in parallel. They just have to require that before you can deliver the results for collect job _n_, the collect jobs for the previous _n-1_ chunks must have been completed.

The problem with this is that we want to allow collectors to skip ahead in the stream of chunks. I suppose this could be accomplished by having the collector send DELETE requests to collect job URIs for those chunks it wants to skip?

Reply:
@tgeoghegan Regarding bullet point 1, what if each report rounds down its timestamp to the nearest n-hour window (a coarse-grained timestamp to protect client privacy, let's say 4 hours for example)? Leader can create a chunk with just enough reports that have timestamp 00:00, doesn't that mean leader will be forced to drop all future reports that mark its timestamp at 00:00?

Reply:
Thanks, everyone, for taking time to help with this. I think we're now at a point where we need to take a step back and think through requirements.

First, in response to feedback from @tgeoghegan and @BranLwyd, not that we have renamed "time-series" to "time-interval" and "fixed-chunk" to "fized-size". Instead of "chunk" we'll just say "batch".

A quick summary to see if we're on the same page. @tgeoghegan points out a desirable property of time-interval tasks: Given the sequence of queries, a Clients know exactly *which* batch their report would be included in, since we have the following properties:
- (1.) a report can belong to exactly one batch
- (2.) batches are non-overlapping
- (3.) a query determines a unique batch 

For "fixed-size" tasks, we currently guarantee something strictly weaker: a Client is know their report is included in at most one batch, but it does not know which one. This is because we have (1.) and (2.) but not (3.).

We don't seem to have a workable solution to this yet:
- (a.) We could have the Client include the batch ID in its report. This is the direct analogue of including the timestamp in the report. However this is a non-starter because it shifts coordination of the batch selection to Clients.
- (b.) We could require that batches don't overlap in time, just as we do for time-interval tasks. @BranLwyd and @junyechen1996 have pointed out various reasons why this is tricky.

These difficulties beg the following essential question: **How important is property (3.)?** Without it, the Leader is capable of sorting reports into batches as it pleases. For example:
* The Leader can batch reports by User-Agent, location, and so on. This is is still possible with (3.), but this would at least force the Leader to "throw away" reports from clients that are not in the target group, but pertain to the query.
* The Leader can do a kind of weak stuffing attack. Stuffing attacks ares still possible with (3.), but again, it would at least force the Leader to throw away any reports that pertain to the query.

Reply:
My take: This threat falls into a big bucket of Sybil-like attacks that I would like to see a more systematic solution for. (Yes Sybil attacks are hard, but they're at least tractable.) Thus I don't think it's worth expending too much effort trying to get (3.). My vote would be to not require it for every query type (and fixed-size queries in particular), but make sure whatever we spell doesn't prevent it.

Reply:
I don't think property (3) is critical for this feature, since leader can do similar things no matter what collection type is used, with pretty trivial effort. If the leader/collector also authors clients, then it can design a task that apply a malicious filter from clients, I don't think this is the type of threat DAP can address on its own.

Reply:
Let's use the open issue to track resolution to this. I'm not convinced that we want to abandon (3) just yet.

Reply:
I think that approach would drop (3.) for time-interval tasks, which at this point I prefer not to do :) Perhaps there is a version of this that'll work.